### PR TITLE
Improve performance of indexer when many attribute options

### LINKED ISSUE #36386: [Issue] Improve performance of indexer when many attribute options
This issue is automatically created based on existing pull request: magento/magento2#36306: Improve performance of indexer when many attribute options

---------
### Description
When there are many product attribute options, the indexer takes a long time to complete. This change improves performance so that the indexing process is much quicker.

I've performed a stand-alone performance analysis between the two approaches here, and I can confirm that with 100,000 values in the `$options` array, the `foreach` approach takes around **5.527919 seconds**, whereas the `array_intersect_key` approach takes around **0.012552** seconds to produce identical results.

The real-world task that triggered this investigation (and resulted in this code change) was on a Magento 2.3.6 website with 10,000 options; reindexing before the change took around 5 minutes, and after the change the same took around 40 seconds.

### Manual testing scenarios

<details><summary>Example speed comparison script</summary>

```php
<?php

// Variables
$rounds = 10;
$optionCount = 100_000;

// Set up
$options = [];
$attributeValues = [];

for ($i = 0; $i < $optionCount; $i++) {
    $options[$i] = [
        'label' => "label $i",
        'value' => $i,
    ];

    $attributeValues[] = $i;
}

// helper for consistently-formatted numbers
function out($label, $value) {
    printf("%s\t%f\n", $label, $value);
}

// ---------------------
echo "\n\nMETHOD 1\n\n";
$timers = [];
for ($i = 0; $i < $rounds; $i++) {
    $attributeLabels = [];

    $start = microtime(true);

    foreach ($options as $option) {
        if (\in_array($option['value'], $attributeValues)) {
            $attributeLabels[] = $option['label'];
        }
    }

    $stop = microtime(true);
    $duration = $stop - $start;
    $timers[] = $duration;

    out($i, $duration);
}
out('avg', array_sum($timers) / count($timers));

// ---------------------
echo "\n\nMETHOD 2\n\n";
$timers = [];
for ($i = 0; $i < $rounds; $i++) {
    $start = microtime(true);

    $options2 = array_intersect_key($options, array_flip($attributeValues));
    $attributeLabels2 = array_column($options2, 'label');

    $stop = microtime(true);
    $duration = $stop - $start;
    $timers[] = $duration;

    out($i, $duration);
}
out('avg', array_sum($timers) / count($timers));


// ---------------------
if ($attributeLabels === $attributeLabels2) {
    echo "\n\nidentical end result\n";
} else {
    echo "\n\ndifferent end result\n";
}
```
</details>

Given there are no functionality changes in this pull request, the website should behave identically before & after the change. The main difference is in the speed of a full reindex when there are product attributes with far too many options.

The unit tests already confirm that these functions behave as expected. (Yes, I've checked that the unit tests cover these functions and behave as expected.)

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [x] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by fredden**
### Description
When there are many product attribute options, the indexer takes a long time to complete. This change improves performance so that the indexing process is much quicker.

I've performed a stand-alone performance analysis between the two approaches here, and I can confirm that with 100,000 values in the `$options` array, the `foreach` approach takes around **5.527919 seconds**, whereas the `array_intersect_key` approach takes around **0.012552** seconds to produce identical results.

The real-world task that triggered this investigation (and resulted in this code change) was on a Magento 2.3.6 website with 10,000 options; reindexing before the change took around 5 minutes, and after the change the same took around 40 seconds.

### Manual testing scenarios

<details><summary>Example speed comparison script</summary>

```php
<?php

// Variables
$rounds = 10;
$optionCount = 100_000;

// Set up
$options = [];
$attributeValues = [];

for ($i = 0; $i < $optionCount; $i++) {
    $options[$i] = [
        'label' => "label $i",
        'value' => $i,
    ];

    $attributeValues[] = $i;
}

// helper for consistently-formatted numbers
function out($label, $value) {
    printf("%s\t%f\n", $label, $value);
}

// ---------------------
echo "\n\nMETHOD 1\n\n";
$timers = [];
for ($i = 0; $i < $rounds; $i++) {
    $attributeLabels = [];

    $start = microtime(true);

    foreach ($options as $option) {
        if (\in_array($option['value'], $attributeValues)) {
            $attributeLabels[] = $option['label'];
        }
    }

    $stop = microtime(true);
    $duration = $stop - $start;
    $timers[] = $duration;

    out($i, $duration);
}
out('avg', array_sum($timers) / count($timers));

// ---------------------
echo "\n\nMETHOD 2\n\n";
$timers = [];
for ($i = 0; $i < $rounds; $i++) {
    $start = microtime(true);

    $options2 = array_intersect_key($options, array_flip($attributeValues));
    $attributeLabels2 = array_column($options2, 'label');

    $stop = microtime(true);
    $duration = $stop - $start;
    $timers[] = $duration;

    out($i, $duration);
}
out('avg', array_sum($timers) / count($timers));


// ---------------------
if ($attributeLabels === $attributeLabels2) {
    echo "\n\nidentical end result\n";
} else {
    echo "\n\ndifferent end result\n";
}
```
</details>

Given there are no functionality changes in this pull request, the website should behave identically before & after the change. The main difference is in the speed of a full reindex when there are product attributes with far too many options.

The unit tests already confirm that these functions behave as expected. (Yes, I've checked that the unit tests cover these functions and behave as expected.)

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [x] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#36386: Improve performance of indexer when many attribute options

### CODE REVIEW

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run Functional Tests B2B

File: app/code/Magento/Elasticsearch/Model/Adapter/BatchDataMapper/ProductDataMapper.php [hunk 1]
```diff
@@ -370,11 +370,8 @@ private function getValuesLabels(Attribute $attribute, array $attributeValues, i
             return $attributeLabels;
         }
 
-        foreach ($options as $option) {
-            if (\in_array($option['value'], $attributeValues)) {
-                $attributeLabels[] = $option['label'];
-            }
-        }
+        $options = array_intersect_key($options, array_flip($attributeValues));
+        $attributeLabels = array_column($options, 'label');
```
Comment (ihor-sviziev) [inline app/code/Magento/Elasticsearch/Model/Adapter/BatchDataMapper/ProductDataMapper.php]:
```suggestion
        $attributeValues = array_flip($attributeValues);
        foreach ($options as $option) {
            if (isset($attributeValues[$option['value']])) {
                $attributeLabels[] = $option['label'];
            }
        }
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
@fredden, the interesting thing is that combination array_flip + foreach + isset does actually a bit better job and keeps the code more readable, understandable, and even memory-efficient.
<details>
  <summary>Code</summary>

```php
<?php

// Variables
$rounds = 10;
$optionCount = 100_000;

// Set up
$options = [];
$attributeValues = [];

for ($i = 0; $i < $optionCount; $i++) {
    $options[$i] = [
        'label' => "label $i",
        'value' => $i,
    ];

    $attributeValues[] = $i;
}

// helper for consistently-formatted numbers
function out($label, $value) {
    printf("%s\t%f\n", $label, $value);
}

$oldMemUsage = memory_get_usage(true);
out('memory usage (MB)', $oldMemUsage / 1024 / 1024);
// ---------------------
echo "\n\nMETHOD 1\n\n";
$timers = [];
for ($i = 0; $i < $rounds; $i++) {
    $attributeLabels = [];

    $start = microtime(true);

    foreach ($options as $option) {
        if (\in_array($option['value'], $attributeValues)) {
            $attributeLabels[] = $option['label'];
        }
    }

    $stop = microtime(true);
    $duration = $stop - $start;
    $timers[] = $duration;

    out($i, $duration);
}
out('avg', array_sum($timers) / count($timers));
$newMemUsage = memory_get_usage(true);
out('memory usage diff (MB)', ($newMemUsage - $oldMemUsage) / 1024 / 1024);
$oldMemUsage = $newMemUsage;

// ---------------------
echo "\n\nMETHOD 2\n\n";
$timers = [];
for ($i = 0; $i < $rounds; $i++) {
    $start = microtime(true);

    $options2 = array_intersect_key($options, array_flip($attributeValues));
    $attributeLabels2 = array_column($options2, 'label');

    $stop = microtime(true);
    $duration = $stop - $start;
    $timers[] = $duration;

    out($i, $duration);
}
out('avg', array_sum($timers) / count($timers));

$newMemUsage = memory_get_usage(true);
out('memory usage diff (MB)', ($newMemUsage - $oldMemUsage) / 1024 / 1024);
$oldMemUsage = $newMemUsage;

// ---------------------
echo "\n\nMETHOD 3\n\n";
$timers = [];
for ($i = 0; $i < $rounds; $i++) {
    $attributeLabels3 = [];

    $start = microtime(true);
    $attributeValues = array_flip($attributeValues);

    foreach ($options as $option) {
        if (isset($attributeValues[$option['value']])) {
            $attributeLabels3[] = $option['label'];
        }
    }

    $stop = microtime(true);
    $duration = $stop - $start;
    $timers[] = $duration;

    out($i, $duration);
}
out('avg', array_sum($timers) / count($timers));

$newMemUsage = memory_get_usage(true);
out('memory usage diff (MB)', ($newMemUsage - $oldMemUsage) / 1024 / 1024);
$oldMemUsage = $newMemUsage;

// ---------------------
if ($attributeLabels === $attributeLabels2 && $attributeLabels === $attributeLabels3) {
    echo "\n\nidentical end result\n";
} else {
    echo "\n\ndifferent end result\n";
}
```
</details>


<details>
  <summary>Result</summary>

```
memory usage (MB)	50.007812


METHOD 1

0	4.841900
1	4.878133
2	4.830100
3	4.889471
4	4.891963
5	4.869839
6	4.850065
7	4.876511
8	4.839044
9	4.837328
avg	4.860435
memory usage diff (MB)	4.003906


METHOD 2

0	0.007926
1	0.008023
2	0.008090
3	0.007927
4	0.008758
5	0.008022
6	0.008395
7	0.008660
8	0.007926
9	0.008017
avg	0.008174
memory usage diff (MB)	8.007812


METHOD 3

0	0.007524
1	0.007231
2	0.007600
3	0.007355
4	0.007706
5	0.007302
6	0.007555
7	0.007458
8	0.007444
9	0.007281
avg	0.007446
memory usage diff (MB)	4.003906


identical end result
```

</details>

File: app/code/Magento/Elasticsearch/Model/Adapter/BatchDataMapper/ProductDataMapper.php (hunk 1 above)
Comment (fredden) [reply app/code/Magento/Elasticsearch/Model/Adapter/BatchDataMapper/ProductDataMapper.php]: Thanks for looking into this and for the suggestion. I've updated the pull request to incorporate this change.

**Comment by fredden** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: app/code/Magento/Elasticsearch/Model/Adapter/BatchDataMapper/ProductDataMapper.php [hunk 2]
```diff
@@ -395,6 +398,14 @@ private function getAttributeOptions(Attribute $attribute, int $storeId): array
...
              */
             $options = $attribute->usesSource() ? $attribute->setStoreId($storeId)->getSource()->getAllOptions() : [];
+
+            if (!empty($options)) {
+                $options = array_combine(
+                    array_column($options, 'value'),
+                    $options
+                );
+            }
+
```
Comment (ihor-sviziev) [inline app/code/Magento/Elasticsearch/Model/Adapter/BatchDataMapper/ProductDataMapper.php]: I believe we should revert this change as well. It doesn't make sense together with foreach
```suggestion
```
Or it does?

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]
✔ Approved.

Failing tests look not related to changes from this PR.

**Comment by engcom-Lima** [comment]
Hi @fredden ,

Thanks for your contribution and collaboration.

I have a query related to this PR. I have created a Magento setup and used Large profile for performance tool . I have almost 1063 attributes and 55 sets of products attribute sets.
When I ran indexer `bin/magento indexer:reindex catalog_product_attribute` it took 22:55 time before PR.
**Below is PR for your reference:**
Screenshot 2022-10-27 at 2 47 50 PM

 After taking this PR I took 15 mins.
**Screenshot:**
Screenshot 2022-10-27 at 2 13 37 PM

According to the above [comment](https://github.com/magento/magento2/pull/36306#issue-1410542721) , indexer should complete less 1 min but it is not happening to my system. Please provide information if I am wrong.

Thanks

**Comment by fredden** [comment]
Thanks for testing this @engcom-Lima. A 7-minute reduction in time seems like success to me. The referenced comment says: "reindexing before the change took around 5 minutes, and after the change the same took around 40 seconds." which is a 4-minute reduction in time; you're seeing better results than this. However, comparing these two results is not fair, because the two websites are very different. The speed difference that you have described is proof that this pull request improves the indexer speed.

**Comment by engcom-Lima** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**
- Install fresh Magento `2.4-develop`

**Manual testing scenario:**

1. Create large amount of data using performance fixtures.
2. Run `bin/magento indexer:reindex catalog_product_attribute` and record the time of completion.

**Before:** :heavy_multiplication_x: 
image

**After:** :heavy_check_mark:  
image

Since it is not affecting functionality no additional manual test case is required.

**Comment by engcom-Hotel** [comment]
Hello,

It seems that the issue is already taken care in this commit: https://github.com/magento/magento2/commit/859e614504da7473dc5023fff94b0c5ffbcf2a9e

Hence we are closing this PR.

Thanks

**Comment by fredden** [comment]
It looks like the commits in this pull request haven't been merged into the `2.4-develop` branch, but the contents of these has been merged in with different commit hashes. @engcom-Hotel please can you help me understand why this pull request wasn't merged via GitHub?

**Comment by engcom-Hotel** [comment]
Hello @fredden,

As seen in the commit link here https://github.com/magento/magento2/commit/859e614504da7473dc5023fff94b0c5ffbcf2a9e, It looks like your changes from this PR got merged itself with the development branch. 

As it is already merged with the mainline branch, we have closed this.

Thanks

**Pull request closed without merging**